elements/howto-accordion/howto-accordion.js

```diff
@@ -0,0 +1,591 @@
+(function() {
+  /**
+   * Define key codes to help with handling keyboard events.
+   */
+  const KEYCODE = {
+    DOWN: 40,
+    LEFT: 37,
+    RIGHT: 39,
+    UP: 38,
+    HOME: 36,
+    END: 35,
+  };
+
+  /**
+   * `HowtoAccordion` is a container element for headings and panels.
+   *
+   * Each heading must be a `<howto-accordion-heading>`. Each panel must be a
+   * `<howto-accordion-panel>` and must be adjacent to a heading.
+   */
+  class HowtoAccordion extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    /**
+     * `connectedCallback` hooks up the even listeners and considers the
+     * `expand`  attribute on the headers to adjust their styling accordingly.
+     */
+    connectedCallback() {
+      // `<howto-accordion-headers>` emit a custom event when the heading is
+      // instructed to expand. The custom event is a nice abstraction because
+      // expansion can be triggered by clicks, keyboard input and attribute
+      // and property changes alike.
+      this.addEventListener('change', this._onChange);
+      // The element also implements roving tab index to switch focus between
+      // the headers. Therefore key presses are intercepted.
+      this.addEventListener('keydown', this._onKeyDown);
+
+      // TODO: Set up MutationObserver to listen for `expand` attribute on the
+      // headings.
+
+      // Wait for `<howto-accordion-heading>` and `<howto-accordion-panel`
+      // to have booted before proceeding.
+      Promise.all([
+        customElements.whenDefined('howto-accordion-heading'),
+        customElements.whenDefined('howto-accordion-panel'),
+      ])
+        .then(_ => {
+        // Acquire all headings inside the element that need to be set up.
+        const headings = this._allHeadings();
+
+        // Give all headings and panels a unique ID. Set up `aria-controls` and
+        // `aria-labelledby` attributes on headings and panels using those IDs.
+        headings.forEach(heading => {
+          // All buttons inside the `HowtoAccordionHeadings` are made
+          // unfocusable here. Only the first heading will be made focusable
+          // afterwards. This is necessary to implement roving tab index.
+          heading._shadowButton.setAttribute('tabindex', -1);
```

Comment:
Don't think we should be exposing our shadow dom internals to the outside world.

If you need to make all the children of a shadow root unfocusable you can put tabindex=-1 on the hosting element.

https://jsbin.com/dadihup/1/edit?html,output